README.md
# aws-mfa-profiles
A package to create credential for AWS using MFA and with the support of multiple profiles

This package wants to simplify the connection to AWS using the MFA supporting also the possibility to use yours profiles.

## Installation
____
You can install the package using `pip install aws-mfa-profiles`.

## Usage
---
There is a little help function callable with `aws-mfa-profiles -h` that it shows this:
```
usage: aws-mfa-profiles [-h] [-p profile] [-t seconds]

Set credential to connect on AWS using MFA

optional arguments:
  -h, --help  show this help message and exit
  -p profile  Profile from which get mfa configuration
  -t seconds  Token expiration time in second from 900 (15 minutes) to 129600 (36 hours) (default: 43200)
```
I set default value for the expiration time session which is `43200` for the expiration time session, this means that the default section is valid for 12 hours. Take in mind that if you use the AWS owner account (root), you can set the session time at maximum at `3600` seconds.

## How does it works?
---
The package read from your `${HOME}/.aws/credentials` file the accounts that you specified and generate a new section in the same file named `${profile_name_selected/choosed}-mfa` with the credentials generated.
Then, you need to define this new account as default for the profiles that you have on `${HOME}/.aws/config`

_credentials_:
```ini
[account_one]
aws_access_key_id = YOUR_ACCESS_KEY_ID
aws_secret_access_key = YOUR_SECRET_ACCESS_KEY
mfa_serial = YOUR_MFA_ARN

[account_two]
aws_access_key_id = YOUR_ACCESS_KEY_ID
aws_secret_access_key = YOUR_SECRET_ACCESS_KEY
mfa_serial = YOUR_MFA_ARN

[account_one-mfa]
aws_access_key_id = YOUR_TEMPORARY_ACCESS_KEY_ID
aws_secret_access_key = YOUR_TEMPORARY_SECRET_ACCESS_KEY
aws_session_token = YOUR_TEMPORARY_SESSION_TOKEN

[account_two-mfa]
aws_access_key_id = YOUR_TEMPORARY_ACCESS_KEY_ID
aws_secret_access_key = YOUR_TEMPORARY_SECRET_ACCESS_KEY
aws_session_token = YOUR_TEMPORARY_SESSION_TOKEN
```

_config_:
```ini
[default]
region = eu-west-1
cli_pager=

[profile profile-that-use-mfa-from-account_one]
role_arn = arn:aws:iam::111111111111:role/your_role
source_profile = account_one-mfa

[profile profile-that-use-mfa-from-account_two]
role_arn = arn:aws:iam::222222222222:role/your_role
source_profile = account_two-mfa
```

CHANGELOG.md
## Changelog
### 2.0.0
- Now, if you run the executable without a specific profile, it starts reading you `credential` file and proposing the profiles that have `mfa_serial` option defined. 
- Added the possibility to use different account with MFA enaled at the same time, using a customized name you can set different `source_profile` in your `config` file.
- Removed the default value for the profile

### 1.1.1
- This version permit to use one single profile at time, running it, it generates a `temporary-credentials-with-mfa` profile in your credential file and it requires you have this temporary profile setted in all profiles in your `config` file.

aws_mfa_profiles/__main__.py
"""Module for AWS MFA credentials management"""
import argparse
import configparser
import logging
from os import environ as os_environ
from sys import exit as sys_exit
import botocore.exceptions
from boto3 import session as boto3_session
from botocore.session import Session as botocore_session
from inquirer import List as inquirer_list, prompt as inquirer_prompt


logging.basicConfig(format='%(levelname)s - %(message)s')
parser = argparse.ArgumentParser(description='Set credential to connect on AWS using MFA')
parser.add_argument('-p', metavar='profile', help='Profile from which get mfa configuration')
parser.add_argument('-t', metavar='seconds', type=int, default=43200, choices=range(900, 129600),
    help='Token expiration time in second from 900 (15 minutes) to 129600 (36 hours) (default: %(default)s)')
sections_list = []
credential_file = configparser.ConfigParser()
try:
    credential_file.read(f'{os_environ["HOME"]}/.aws/credentials')
except configparser.Error:
    logging.error('%s/.aws/credentials', os_environ['HOME'])
    sys_exit(1)


def check_profile_exists(profile: str):
    """Function to check if the provided profile exists"""
    if not credential_file.has_section(profile):
        logging.error('%s doesn\'t exists in your credentials file', profile)
        sys_exit(1)


def check_profile_mfa(profile: str):
    """"Function to check if the provided profile has the mfa"""
    if not credential_file.has_option(profile, 'mfa_serial'):
        logging.error('%s haven\'t got the mfa_serial defined in the credential file', profile)
        sys_exit(1)


def check_token_validation(prompt: str):
    """Function to check if the provided tocken is valid"""
    if prompt.isdecimal() and 6 <= len(prompt):
        logging.info('Token has at least 6 characters, this is ok')
        logging.info('Token is only numerical, this is ok')
    else:
        logging.error('Token is invalid!')
        sys_exit(1)


def generate_credentials(profile: str, aws_token_validity: int, token: str):
    """Function to generate the credentials with MFA"""
    mfa_session = botocore_session(profile=profile)
    mfa_serial = mfa_session.get_scoped_config().get('mfa_serial')
    client = generate_client_session(profile)
    try:
        response = client.get_session_token(
            DurationSeconds=aws_token_validity,
            SerialNumber=mfa_serial,
            TokenCode=token
        )
        return response
    except botocore.exceptions.ClientError as error:
        logging.error('Ops, you encounter an exception: %s', error.response['Error']['Code'])
        return None


def generate_client_session(profile: str):
    """Function to create a client session"""
    session = boto3_session.Session(
        profile_name=profile
    )
    client = session.client('sts')
    return client


def read_profiles():
    """Function to read the valid profile in the credentials file"""
    try:
        sections = credential_file.sections()
        for section in sections:
            if '-mfa' in section:
                continue
            if not credential_file.has_option(section, 'mfa_serial'):
                continue
            sections_list.append(section)
    except configparser.Error:
        logging.error('Something goes wrong during the read_profiles function')


def set_aws_variables(profile: str, credentials):
    """Function to add the temporary credential to the credentials file"""
    section_name = f'{profile}-mfa'
    try:
        credential_file.remove_section(section_name)
    except configparser.NoSectionError:
        logging.warning('No section %s to delete ', section_name)
    credential_file.add_section(section_name)
    credential_file[section_name]['aws_access_key_id'] = credentials['Credentials']['AccessKeyId']
    credential_file[section_name]['aws_secret_access_key'] = credentials['Credentials']['SecretAccessKey']
    credential_file[section_name]['aws_session_token'] = credentials['Credentials']['SessionToken']
    with open(f'{os_environ["HOME"]}/.aws/credentials', encoding="utf-8", mode='w') as configfile:
        credential_file.write(configfile)


def main():
    """Main function"""
    args = parser.parse_args()
    profile = args.p
    aws_token_validity = args.t
    if not profile:
        read_profiles()
        profiles_list = [inquirer_list('profile_name',
            message='Choose the profile from which create temporary MFA credentials',
            choices=sections_list)]
        selected_profile = inquirer_prompt(profiles_list)
        if not selected_profile:
            sys_exit(1)
        profile = selected_profile['profile_name']
    check_profile_exists(profile)
    check_profile_mfa(profile)
    token = input("Insert MFA token: ")
    check_token_validation(token)
    response = generate_credentials(profile, aws_token_validity, token)
    if not response:
        sys_exit(1)
    set_aws_variables(profile, response)


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        pass
